### PR TITLE
FileStorageService: verify path resolves under the store directory

### PR DESCRIPTION
**Pull request by tmds**

### CODE REVIEW

File: src/BaGet.Core/Services/FileStorageService.cs [hunk 1]
```diff
@@ -96,13 +99,19 @@ public Task DeleteAsync(string path, CancellationToken cancellationToken = defau
...
             {
                 throw new ArgumentException("Path is required", nameof(path));
             }
 
-            return Path.Combine(_storePath, path);
+            // Verify path is under the _storePath.
+            string fullPath = Path.GetFullPath(Path.Combine(_storePath, path));
+            if (!fullPath.StartsWith(_storePath))
+            {
+                throw new ArgumentException("Path resolves outside store path");
```
Comment (loic-sharma) [inline src/BaGet.Core/Services/FileStorageService.cs]: Could you add a `nameof(path)` to be consistent with line 104?

Comment (loic-sharma) [reply src/BaGet.Core/Services/FileStorageService.cs]: Also please add a unit test in [`FileStorageServiceTests.cs`](https://github.com/loic-sharma/BaGet/blob/master/tests/BaGet.Core.Tests/Services/FileStorageServiceTests.cs)

**Comment by loic-sharma** [comment]
Argh. It looks like the CI is broken for external contributors. I'll fix that tonight 😓

File: tests/BaGet.Core.Tests/Services/FileStorageServiceTests.cs [hunk 2]
```diff
@@ -191,6 +234,25 @@ protected async Task<string> ToStringAsync(Stream input)
...
+                get
+                {
+                    string fullPath = Path.GetFullPath(_storePath);
+                    yield return "../file";
+                    yield return ".";
+                    yield return $"../{Path.GetFileName(_storePath)}";
+                    yield return $"../{Path.GetFileName(_storePath)}suffix";
+                    yield return $"../{Path.GetFileName(_storePath)}suffix/file";
+                    yield return fullPath;
+                    yield return fullPath + Path.DirectorySeparatorChar;
```
Comment (loic-sharma) [inline tests/BaGet.Core.Tests/Services/FileStorageServiceTests.cs]: This test case fails on Windows. See my comment on line 109 of `FileStorageService`.

File: src/BaGet.Core/Services/FileStorageService.cs [hunk 3]
```diff
@@ -96,13 +99,19 @@ public Task DeleteAsync(string path, CancellationToken cancellationToken = defau
...
-            // TODO: This should check that the result is in _storePath for security.
             if (string.IsNullOrEmpty(path))
             {
                 throw new ArgumentException("Path is required", nameof(path));
             }
 
-            return Path.Combine(_storePath, path);
+            // Verify path is under the _storePath.
+            string fullPath = Path.GetFullPath(Path.Combine(_storePath, path));
+            if (!fullPath.StartsWith(_storePath))
```
Comment (loic-sharma) [inline src/BaGet.Core/Services/FileStorageService.cs]: If the `fullPath` and `_storePath` are identical, this check will pass. This causes the test `fullPath + Path.DirectorySeparatorChar;` to fail. I would add something like:

```
if (fullPath.Equals(_storePath, StringComparison.InvariantCultureIgnoreCase))
{
    throw new ArgumentException("Path must be a file within the store path", nameof(path));
}
```

File: tests/BaGet.Core.Tests/Services/FileStorageServiceTests.cs (hunk 2 above)
Comment (loic-sharma) [reply tests/BaGet.Core.Tests/Services/FileStorageServiceTests.cs]: I'm loving how thorough these tests are 👍

File: src/BaGet.Core/Services/FileStorageService.cs (hunk 3 above)
Comment (tmds) [reply src/BaGet.Core/Services/FileStorageService.cs]: I've added a length check. I also changed the startswith compare to an Ordinal one.

**Comment by loic-sharma** [comment]
Thank you! :)

**Pull request merged**